### ISSUE TITLE
Startbon Fehler - TIPP

### ISSUE THREAD

**Issue by shawtaylor** [a-sit-plus/at-registrierkassen-mustercode#484]
Hallo,

ich bekomme immer eine Fehlermeldung beim Scannen mit den BMF APP.

Allgemeiner Fehler
Belegprüfung fehlgeschlagen
Bitte überfrüfen Sie den Beleg.

_R1-AT1_Kasse-00-50-56-0-00-01_4015_2017-03-24T11:42:22
_0,00_0,00_0,00_0,00_0.00
_SlGnRCR3A/eXTQ4Q3XLiVQ==_4E0E0D46
_NfsjRate+Pexc4OCpgrA8tyYJYnlA3gphzFfW8XEiIsQquOQSDbv0qVTdmhGgqvgbm6GWVmUKuX2SpWKOOqQqQ==

Ich glaube das es am letzten Element *Signatur* liegt. Einer eine Idee wie ich die richtige Signatur bekommen?

![meinbild28](https://cloud.githubusercontent.com/assets/13533887/24300300/494a373a-10ac-11e7-8bbf-1fa1202d7594.jpeg)

**Comment by ErichFreitag** [comment]
Es fehlt der Eintrag für die Belegverkettung.
Im Steuersatz Besonders findet sich ein "." statt ",".
Wenn die Signatur fehlerhaft ist bekommen sie das auch extra ausgewiesen.

**Comment by shawtaylor** [comment]
***Signaturerstellung        
*** Muss ich hier etwss bestimmtes (textString ) übergeben? 
*** Verstehe ich nicht
 
  string textString = "Das ist ein Beispieltext der anstelle eines Registrierkasseneintrags signiert wird... ";
            byte[] text = Encoding.Unicode.GetBytes(textString);
            SHA256Managed crypt = new SHA256Managed();
            byte[] hash = crypt.ComputeHash(text, 0, text.Length);
            crypt.Dispose();
            
            string hashString = Convert.ToBase64String(hash);
       // String pin = "123456";
           string pin = CardINI.ReadString("Card", "PIN", "123456");

            byte[] signature = null;
            for (int i = 0; i < 4; i++) {
                Stopwatch sw = new Stopwatch();
                sw.Restart();
                signature = smartCard.sign(**pin, hash**);
                sw.Stop();
            
            }

            smartCard.prepareSignature();
            for (int i = 0; i < 4; i++) {
                Stopwatch sw = new Stopwatch();
                sw.Restart();
                signature = smartCard.signWithoutSelection(pin, hash);
                sw.Stop();
             
            }
            string s3 = Convert.ToBase64String(signature);
            Console.WriteLine("Signature:  " + s3);

            CardINI.Write("Card", "Signature", s3.ToString());
            CardINI.Write("Card", "SignatureTime", DateTime.Now.ToString() );

            bool bOK = smartCard.verify(signature, text);

//***********************************************************************************************


           // Console.WriteLine("SHA256 Hash: " + hashString);

            string szBeleg = CardINI.ReadString("Card", "BelegNr", "1");
            Console.WriteLine("BelegNr: " + szBeleg.ToString());
            
            
           // CardINI.Write("Card", "SHA256 Hash", hashString.ToString());
        
         
            long fUmsatz = CardINI.ReadLong("Card", "UmsatzRohwert", 0);
            Console.WriteLine("UmsatzRohwert: " + fUmsatz.ToString());

            long fUmsatzZeahler = CardINI.ReadLong("Card", "fUmsatzZeahler", 0);
            Console.WriteLine("fUmsatzZeahler: " + fUmsatzZeahler.ToString());

            
            string szKassenID = CardINI.ReadString("Signi", "Kassenidentifikationsnummer", "Kasse1");
            Console.WriteLine("Kasse: " + szKassenID.ToString());
            string szKey = CardINI.ReadString("Signi", "EditAESKey", "");
            Console.WriteLine("Key: " + szKey.ToString());
            byte[] aesBytes = Convert.FromBase64String(szKey);

            byte[] encryptedBytes = Registrierkasse.RKAesIcm.Encrypt(fUmsatz + fUmsatzZeahler, szKassenID, szBeleg, aesBytes);
            Console.WriteLine("Verschlüsselter Umsatz (B64): " + Convert.ToBase64String(encryptedBytes));
            CardINI.Write("Card", "UmsatzKey256", Convert.ToBase64String(encryptedBytes));

            long  lWert = Registrierkasse.RKAesIcm.Decrypt(encryptedBytes, szKassenID, szBeleg, aesBytes);
            CardINI.Write("Card", "UmsatzKey256", Convert.ToBase64String(encryptedBytes));

            Console.WriteLine("Verschlüsselter Umsatz (B64) Decrypt: " + lWert.ToString());

**Comment by perler-at** [comment]
Hallo!

Ich bekomme den selben Fehler und sehe den Wald vor lauter Bäumen nicht mehr.

QR-Code im Klartext:
_R1-AT1_33350458_1_2017-03-28T15:41:48_0,00_0,00_0,00_0,00_0,00_DkK9gB/km/k=_659E8914_UFwiaB4+1HA=_6rjhnoFtcpZJqlGyZA1u8SKHS6E7bdoINn_zn2m9PY-k4_YDIjGB4iSDDxtIhgQ5Aijdi6E9WD-sE3-O7uB5ig

Ich würde mich über einen Tipp sehr freuen.

![qr](https://cloud.githubusercontent.com/assets/20278180/24458378/0d85d998-1499-11e7-98ae-9c07ee9f2e7c.jpg)

**Comment by ztp-mino** [comment]
da sind unterstriche in der signatur

**Comment by perler-at** [comment]
Danke für den Hinweis.
Die base64url Decodierung der Signatur hat gefehlt.
Nun erhalte ich schon mal ein netagives Prüfergebnis mit dem Hinweis in FON "Der Signaturwert enthält ungültige Zeichen für die BASE64-Kodierung."

Der akeulle Code:
`_R1-AT1_33350458_1_2017-03-28T15:41:48_0,00_0,00_0,00_0,00_0,00_DkK9gB/km/k=_659E8914_UFwiaB4+1HA=_6Gu9T444xvAs0vYuqMZFegGC6AsPb66dvHiRkYkJT1XXCl+qT/u7A8hh3fZXpIC9U3uxyTHMRgL3 wZpaw4Nz9A==`

**Comment by lebail** [comment]
Ja, Leerzeichen darf da keines drinnen sein.  Wenn das wirklich so herauskommt, hat Ihre Base64-Implementierung einen Fehler.

**Comment by DavStock** [comment]
Und was würden Sie zu diesem QR-Code (für Startbeleg) sagen?

![14493_2016](https://cloud.githubusercontent.com/assets/26740817/24504466/b6906c48-1556-11e7-91a3-f03fc965dd7f.jpg)

Laut Prüftool ist die kryptographie der Signatur fehlerhaft. Klartext des QR-Codes ist:
`_R1-AT1_KRO-CTG-WSRKA_271_2017-03-30T13:16:07_0,00_0,00_0,00_0,00_0,00_fQeXmFQxrUJZ8OdM87aKsw==_7F4D87C4_S1JPLUNURy0=_gbdnDAhIKGAY5yVtMQbINEb3Ha/trjKdlmHkH4K1vcO7JISW2q3tWVXQCr7gafQPZ0UT427KXRly6jmowY7rqA==`

Kann es sein, dass die Base64Url des Umsatzzählers zu lange ist? Kann dadurch ein kryptographischer Fehler der Signatur entstehen oder hat diese Fehlermeldung ausschließlich etwas mit der JWS Formatierung und Signierung zu tun?

**Comment by ErichFreitag** [comment]
Da das was zu signieren ist gehasht wird hat es mit dem Inhalt nichts zu tun.

**Comment by DavStock** [comment]
Das heisst jetzt was? Kryptographischer Fehler bezieht sich auf den gesamten Klartext (siehe oben) oder nur auf den letzten gehashten Teil (JWS Repräsentation in Base64)?

**Comment by ErichFreitag** [comment]
Das was sie signieren - also header.payload gehasht.

**Comment by DavStock** [comment]
hm...ok - dann werde ich wohl da weiter tüfteln müssen! Danke @ErichFreitag für den wiederholten Einsatz!!!
